experimental/Schemes/WeilDivisor.jl

```diff
@@ -0,0 +1,352 @@
+    # TODO: Do we want to require that the different effective divisors 
+    # have the same underlying covering? Probably not.
+    if check
+      for D in keys(coefficients)
+        space(D) === X || error("component of divisor does not lay in the given scheme")
+        parent(coefficients[D]) === R || error("coefficients do not lay in the given ring")
+      end
+    end
+    return new{typeof(X), CoefficientRingType, CoefficientRingElemType}(X, R, coefficients)
+  end
+end
+
+### type getters 
+scheme_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = S
+scheme_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = S
+coefficient_ring_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = U
+coefficient_ring_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = U
+coefficient_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = V
+coefficient_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = V
+
+### getter methods
+@Markdown.doc """
+    scheme(D::WeilDivisor)
+
+Return the `CoveredScheme` ``X`` on which `D` is defined.
+"""
+scheme(D::WeilDivisor) = D.X
+
+getindex(D::WeilDivisor, I::IdealSheaf) = (D.coefficients)[I]
+
+@Markdown.doc """
+    components(D::WeilDivisor)
+
+Return the irreducible components ``Eⱼ`` of the divisor 
+``D = Σⱼ aⱼ ⋅ Eⱼ``.
+"""
+components(D::WeilDivisor) = [ Z for Z in keys(D.coefficients)]
+coefficient_dict(D::WeilDivisor) = D.coefficients
+coefficient_ring(D::WeilDivisor) = D.R
+
+set_name!(X::WeilDivisor, name::String) = set_attribute!(X, :name, name)
+name_of(X::WeilDivisor) = get_attribute(X, :name)::String
+has_name(X::WeilDivisor) = has_attribute(X, :name)
+
+function setindex!(D::WeilDivisor, c::RingElem, I::IdealSheaf)
+  parent(c) === coefficient_ring(D) || error("coefficient does not belong to the correct ring")
+  coefficient_dict(D)[I] = c
+end
+
+@Markdown.doc """
+    WeilDivisor(X::CoveredScheme, R::Ring)
+
+Return the zero `WeilDivisor` over `X` with coefficients 
+in `R`.
+"""
+function WeilDivisor(X::CoveredScheme, R::Ring)
+  D = IdDict{IdealSheaf, elem_type(R)}()
+  return WeilDivisor(X, R, D)
+end
+
+@Markdown.doc """
+    WeilDivisor(I::IdealSheaf, R::Ring)
+
+Return the `WeilDivisor` ``D = 1 ⋅ V(I)`` with coefficients 
+in ``R`` for a sheaf of prime ideals ``I``.
+"""
+function WeilDivisor(I::IdealSheaf, R::Ring)
+  D = WeilDivisor(space(I), R)
+  D[I] = one(R)
+  return D
+end
+
+@Markdown.doc """
+    WeilDivisor(I::IdealSheaf)
+
+Return the `WeilDivisor` ``D = 1 ⋅ V(I)`` with coefficients
+in ``ℤ`` for a sheaf of prime ideals ``I``.
+"""
+function WeilDivisor(I::IdealSheaf)
+  D = WeilDivisor(space(I), ZZ)
+  D[I] = one(ZZ)
+  return D
+end
+
+### copy constructor
+function copy(D::WeilDivisor) 
+  new_dict = IdDict{IdealSheaf, elem_type(coefficient_ring_type(D))}()
+  for I in keys(coefficient_dict(D))
+    new_dict[I] = D[I]
+  end
+  return WeilDivisor(scheme(D), coefficient_ring(D), new_dict)
+end
+
+function Base.show(io::IO, D::WeilDivisor)
```

Comment:
The show function can have context
E.g. `IOContext(io, :compact => true)`
for better printing in arrays for instance. Please provide it. (Just grep the source for examples on how to do this)